compiler/lib/Dialect/mhlo/Transforms/CanonicalizeExt.cpp

```diff
@@ -1377,6 +1377,47 @@ struct FoldLargeConcatenate : public OpRewritePattern<mhlo::ConcatenateOp> {
   }
 };
 
+struct CanonicalizeClamp : public OpRewritePattern<mhlo::ClampOp> {
+  using OpRewritePattern<mhlo::ClampOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(mhlo::ClampOp op,
+                                PatternRewriter &rewriter) const override {
+    auto minConst = op.getMin().getDefiningOp<mhlo::ConstantOp>();
+    auto maxConst = op.getMax().getDefiningOp<mhlo::ConstantOp>();
+    if (!minConst || !maxConst) {
+      return failure();
+    }
+    auto minAttr =
+        dyn_cast_if_present<DenseElementsAttr>(minConst.getValueAttr());
+    auto maxAttr =
+        dyn_cast_if_present<DenseElementsAttr>(maxConst.getValueAttr());
+    if (!minAttr || !maxAttr) {
+      return failure();
+    }
+    if (!minAttr.isSplat() || !maxAttr.isSplat()) {
+      return failure();
+    }
+    // replace splat const with scalar
+    if (minAttr.getType().getRank() > 0 && maxAttr.getType().getRank() > 0) {
+      minAttr = minAttr.dyn_cast<SplatElementsAttr>().resizeSplat(
+          minAttr.getType().clone(SmallVector<int64_t>{}));
+      maxAttr = maxAttr.dyn_cast<SplatElementsAttr>().resizeSplat(
+          maxAttr.getType().clone(SmallVector<int64_t>{}));
+      minConst = rewriter.create<mhlo::ConstantOp>(minConst.getLoc(), minAttr);
+      op->setOperand(0, minConst);
+      maxConst = rewriter.create<mhlo::ConstantOp>(maxConst.getLoc(), maxAttr);
+      op->setOperand(2, maxConst);
+      return success();
+    }
+    // remove op if min/max are out of range
+    if (minAttr.getSplatValue<FloatAttr>().getValue().isNegInfinity() &&
```

Comment:
Should check dtype is float? Since it will core when it is IntergerAttr.

Reply:
(BTW, if you want to add Interger support, I think `clamp(INT_MIN, x, INT_MAX) => clamp` is acceptable.